op-challenger/game/fault/trace/asterisc/provider.go

```diff
@@ -0,0 +1,256 @@
+package asterisc
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/ethereum-optimism/optimism/op-challenger/config"
+	"github.com/ethereum-optimism/optimism/op-challenger/game/fault/trace/cannon"
+	"github.com/ethereum-optimism/optimism/op-challenger/game/fault/types"
+	preimage "github.com/ethereum-optimism/optimism/op-preimage"
+	"github.com/ethereum-optimism/optimism/op-program/host/kvstore"
+	"github.com/ethereum-optimism/optimism/op-service/ioutil"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/log"
+)
+
+const (
+	proofsDir      = "proofs"
+	diskStateCache = "state.json.gz"
+)
+
+type AsteriscMetricer interface {
+	RecordAsteriscExecutionTime(t float64)
+}
+
+type AsteriscTraceProvider struct {
+	logger         log.Logger
+	dir            string
+	prestate       string
+	generator      cannon.ProofGenerator
+	gameDepth      types.Depth
+	preimageLoader *cannon.PreimageLoader
+
+	// lastStep stores the last step in the actual trace if known. 0 indicates unknown.
+	// Cached as an optimisation to avoid repeatedly attempting to execute beyond the end of the trace.
+	lastStep uint64
+}
+
+func NewTraceProvider(logger log.Logger, m AsteriscMetricer, cfg *config.Config, localInputs cannon.LocalGameInputs, dir string, gameDepth types.Depth) *AsteriscTraceProvider {
+	return &AsteriscTraceProvider{
+		logger:         logger,
+		dir:            dir,
+		prestate:       cfg.AsteriscAbsolutePreState,
+		generator:      NewExecutor(logger, m, cfg, localInputs),
+		gameDepth:      gameDepth,
+		preimageLoader: cannon.NewPreimageLoader(kvstore.NewDiskKV(cannon.PreimageDir(dir)).Get),
+	}
+}
+
+func (p *AsteriscTraceProvider) SetMaxDepth(gameDepth types.Depth) {
+	p.gameDepth = gameDepth
+}
+
+func (p *AsteriscTraceProvider) Get(ctx context.Context, pos types.Position) (common.Hash, error) {
+	traceIndex := pos.TraceIndex(p.gameDepth)
+	if !traceIndex.IsUint64() {
+		return common.Hash{}, errors.New("trace index out of bounds")
+	}
+	proof, err := p.loadProof(ctx, traceIndex.Uint64())
```

Comment:
We'll need to expand this to a `uint128` to support the expanded trace index bounds. Will follow up sync in our meeting today.

Reply:
I understood that we do not have to expand this, because traceIndex at this point is already safe, not overflowing over Uint64. There is no cannon trace provider fix as well.